### ISSUE TITLE
Dynamo locks should have a configurable length

### ISSUE THREAD

**Issue by alexwlchan** [wellcomecollection/platform#5211]
In the current implementation of scala-libs, the DynamoLockDao doesn't have a configurable expiry time – it always uses the default, which is 3 minutes. See https://github.com/wellcomecollection/scala-libs/blob/main/storage_typesafe/src/main/scala/uk/ac/wellcome/storage/typesafe/DynamoLockDaoBuilder.scala#L18-L20

This means that all locks will expire 3 minutes after they were created, even if the underlying application hasn't finished doing whatever it was locking around.

In theory, this could allow multiple services to access the same locked resource simultaneously:

1. Service #​1 picks up a new message from SQS. There is no existing lock, so it creates one and starts working on the task.
2. After 3 minutes, lock #​1 expires.
3. Sometime after that, the SQS queue visibility timeout expires. The message is now available for another service.
4. Service #​2 picks up the same message from SQS. Although service #​1 is still working, its lock has expired, so #​2 creates a new lock and starts working on the task.
 
<img width="719" alt="Screenshot 2021-06-25 at 07 41 13" src="https://user-images.githubusercontent.com/301220/123381489-b8f0f600-d588-11eb-9cc1-625115791562.png">

We should make the lock timeout configurable.

## The bag versioner

Where this is causing us visible issues is in the bag versioner: for some reason lock #​1 isn't being cleaned up properly, and the SQS visibility timeout is shorter than the expiry time. When we retry the SQS message, we get blocked by the original lock:

<img width="444" alt="Screenshot 2021-06-25 at 07 46 28" src="https://user-images.githubusercontent.com/301220/123382097-74b22580-d589-11eb-8e96-b35933800b66.png">

Making the lock timeout configurable will allow us to reduce the expiry time and avoid this issue.

## The bag replicator

I don't think this actually causes issues in the replicator, but it's certainly not the expected behaviour. If you'd asked me to describe how the replicator works, I'd get it wrong.

* Expected behaviour: only one replicator can be writing to a prefix at a time, unless it's a very long replication (8 hours+, or whatever the SQS visibility timeout is)

* Actual behaviour: multiple replicators can be writing to a prefix at the same time. I don't think this is a big issue because of other protections in the replicator, but we're not as well protected as we might have thought. In particular:

     - Replicator #​1 picks up a new message from SQS. There's nothing in the destination prefix, so it uses "write-immediately" replication – it will write objects directly into the destination prefix without checking if there's something there.
     - Replicator #​2 picks up the same message from SQS, say if the bag is very large. There are objects in the destination prefix, so it uses "read-before-write" replication – before writing an object, it inspects the existing object and compares the two.

    Possible failures and why I think they're not an issue:

    - If #​2 writes an object before #​1, then #​1 will merrily overwrite it without checking what was already there. This shouldn't be an issue because the two replicators should work through objects in the same order, so #​2 will never catch up.
    - If #​1 hasn't written an object before #​2 starts, then #​2 will ignore what was already there. Because we have an 8 hour visibility timeout on the queue, that seems pretty unlikely. We could guard against that by getting #​1 to work through objects in size order, smallest first – but that seems like overkill

## tl;dr

I've convinced myself that this is probably fine and doesn't need further investigation – I'll fix the bug and we can call it done.

**Comment by alexwlchan** [comment]
Closing as done, and it turns out this wasn't actually the bug causing the storage service issues – we were actually failing to release some locks correctly.
